lib/realtime/vehicle.ex

```diff
@@ -62,20 +77,119 @@ defmodule Realtime.Vehicle do
       }
     }
   """
-  @spec decode(term()) :: t()
-  def decode(%{} = json) do
+  @spec decode(term(), keyword) :: t()
+  def decode(%{} = json, opts \\ []) do
+    _opts = Keyword.merge(@default_opts, opts)
+    stop_times_on_route_fn = Keyword.get(opts, :stop_times_on_route_fn)
+
+    id = json["id"]
+    label = json["vehicle"]["vehicle"]["label"]
+    timestamp = json["vehicle"]["timestamp"]
+    direction_id = json["vehicle"]["trip"]["direction_id"]
+    route_id = json["vehicle"]["trip"]["route_id"]
+    trip_id = json["vehicle"]["trip"]["trip_id"]
+    current_stop_status = decode_current_status(json["vehicle"]["current_status"])
+    stop_id = json["vehicle"]["stop_id"]
+
+    stop_times_on_route =
+      try do
+        stop_times_on_route_fn.(route_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    ordered_stop_times =
+      if direction_id == 0, do: Enum.reverse(stop_times_on_route), else: stop_times_on_route
+
+    {percent_of_the_way_to_next_timepoint, next_timepoint_stop_time} =
+      percent_of_the_way_to_next_timepoint(ordered_stop_times, stop_id)
+
     %__MODULE__{
-      id: json["id"],
-      label: json["vehicle"]["vehicle"]["label"],
-      timestamp: json["vehicle"]["timestamp"],
-      direction_id: json["vehicle"]["trip"]["direction_id"],
-      route_id: json["vehicle"]["trip"]["route_id"],
-      trip_id: json["vehicle"]["trip"]["trip_id"],
-      current_status: decode_current_status(json["vehicle"]["current_status"]),
-      stop_id: json["vehicle"]["stop_id"]
+      id: id,
+      label: label,
+      timestamp: timestamp,
+      direction_id: direction_id,
+      route_id: route_id,
+      trip_id: trip_id,
+      current_stop_status: current_stop_status,
+      stop_id: stop_id,
+      current_timepoint_status:
+        timepoint_status(current_stop_status, next_timepoint_stop_time, stop_id),
+      timepoint_id: next_timepoint_stop_time && next_timepoint_stop_time.timepoint_id,
+      percent_of_the_way_to_timepoint: percent_of_the_way_to_next_timepoint
+    }
+  end
+
+  @spec percent_of_the_way_to_next_timepoint([StopTime.t()], Stop.id()) :: {
+          non_neg_integer(),
+          StopTime.t() | nil
+        }
+  def percent_of_the_way_to_next_timepoint([], _stop_id), do: {0, nil}
+
+  def percent_of_the_way_to_next_timepoint(stop_times, stop_id) do
+    {past_stop_times, current_stop_time, future_stop_times} =
+      split_stop_times(stop_times, stop_id)
+
+    next_timepoint_stop_time =
+      Enum.find([current_stop_time | future_stop_times], &is_a_timepoint?(&1))
+
+    {
+      percent_of_the_way(
+        count_to_timepoint(past_stop_times),
+        count_to_timepoint(future_stop_times)
+      ),
+      next_timepoint_stop_time
     }
   end
 
+  @spec split_stop_times([StopTime.t()], Stop.id()) ::
+          {[StopTime.t()], StopTime.t() | nil, [StopTime.t()]}
+  defp split_stop_times(stop_times, stop_id) do
```

Comment:
This looks like it would be more efficient w/ `Enum.split_with/2`: https://hexdocs.pm/elixir/Enum.html#split_with/2